### ISSUE TITLE
Any way to specify that an array property should be serialized without a property key?

### ISSUE THREAD

**Issue by mrhappyasthma** [StringEpsilon/PhpSerializerNET#32]
I'm writing an application that needs to be compatible with a legacy client whose code I can't change.

I have an issue where I have an object that has a nested array. The client expects the array to be serialized without a string key. It's just indexed starting with `i:0`.

E.g.

My class:

```
public class MyClass
{
     // How to avoid the `objects` key prefix?
    public List<Objects>? objects { get; set; }

    [PhpProperty("some_string")]
    public string? some_string{ get; set; }

    [PhpProperty("some_string2")]
    public string? some_string2{ get; set; }

    [PhpProperty("some_string3")]
    public string? some_string3{ get; set; }

    [PhpProperty("some_string4")]
    public string? some_string4{ get; set; }
}
```

another option I tried was just using an object but not a list:

```
     // How to have this be an integer index?
    public Objects? objects { get; set; }
```

The output I'm hoping to match would look like this:

```
// Notice the array index `0` is used without any string key
a:5:{i:0;a:2:
```

What I'm actually getting [List version]:

```
a:5:{s:7:"objects";a:1:{i:0;a:2:
```

What I'm getting [without List version]:

```
a:5:{s:7:"objects";a:2:
```

Thanks in advance.

**Comment by StringEpsilon** [comment]
I'm not quite sure I understand the requirements.

What I'm reading is you have a structure like this:

```
MyClass
    -> 0:List0
        -> 0:"list0-foo"
        -> 1:"list0-bar"
        -> 2:"list0-baz"
        ...
    -> 1:List2
        -> 0:"list1-foo"
        -> 1:"list1-bar"
        -> 2:"list1-baz"
        ...
```

If that's the case, you can just work with a list of lists:

```cs
var data = new List<List<string>>(){
    new List<string>() {
        "list1-foo",
        "list1-bar",
        "list1-baz",
    },
    new List<string>() {
        "list1-foo",
        "list1-bar",
        "list1-baz",
    },
};

System.Console.WriteLine(PhpSerialization.Serialize(data));
```

outputs:

```
 a:2:{i:0;a:3:{i:0;s:9:"list1-foo";i:1;s:9:"list1-bar";i:2;s:9:"list1-baz";}i:1;a:3:{i:0;s:9:"list1-foo";i:1;s:9:"list1-bar";i:2;s:9:"list1-baz";}}
``` 

There is currently no way to serialize an object into an integer-keyed array. I never considered that, to be honest. But I guess I could add that feature.

**Comment by StringEpsilon** [comment]
I cobbled together a quick version:

https://github.com/StringEpsilon/PhpSerializerNET/blob/f71c31b863d4c8901eeed8bd4eaa0808586f1630/PhpSerializerNET.Test/Serialize/ArraySerialization.cs#L24 

Would that be useful?

**Comment by mrhappyasthma** [comment]
I appreciate the responses!

The `useIntegerKeys` is very close to what I think we need. But I actually want the integer key applying to the parent object in my instance.

Also I do not need deserialization of this format. In all cases I've seen for my needs, the key order is fixed. I'm not sure if that's a strict requirement. I'd have to play around with it but I can double check.

Essentially I'm trying to write an application that serializes data exactly how an existing program (which is closed source) consumes it. But I know the format due to packet sniffing.

It seems they do something a bit unconventional. I imagine it's like a dictionary that supports both integer keys (for array entries) but also string keys (for other top level fields).

(Also let me know if there is some better way to accomplish this, but it seems like it may need adjustments.)

---------

So I have something like this:

```
a:4:{i:0;a:3:{s:1:"a";s:5:"value";s:1:"b";s:5:"value";s:1:"c";s:5:"value";}i:1;a:3:{s:1:"a";s:5:"value";s:1:"b";s:5:"value";s:1:"c";s:5:"value";}s:1:"d";s:5:"value";s:1:"e";s:5:"value";}
```

Or better visually represented:

```
Array
(
    [0] => Array
        (
            [a] => value
            [b] => value
            [c] => value
        )
     [1] => Array
        (
            [a] => value
            [b] => value
            [c] => value
        )

    [d] => value
    [e] => value
)
```

But the keys in the first array, `0`, and `1` are ints. And `d` and `e` are strings.

So I think I don't want ` [PhpSerializeArray(useIntegerKeys: true)] ` applied to my inner object (as I want `a`, `b`, and `c`, to be string keys). But on my parent, wrapper object I want the key to be used.

Something sort of like this:

```
public class MyClass
{
    [PhpSerializeArray(useIntegerKeys: true)]   // or maybe [PhpProperty(userIntegerKeys:true)]
    public List<Objects>? innerArray { get; set; }

    [PhpProperty("d")]
    public string? d { get; set; }

    [PhpProperty("e")]
    public string? e { get; set; }
}
```

**Comment by StringEpsilon** [comment]
I'll have to think about that. I wasn't even aware that mixing integer keys with string keys is allowed until now.

In the mean time, you can use a dictionary:

```cs
object value = new InnerObject(); // just for example.

var data = new Dictionary<object, object>() {
    {1, new List<object>() { value, value, value },
    {2, new List<object>() { value, value, value },
    {3, new List<object>() { value, value, value },
    {"a", "A"},
    {"b", "B"},
    {"c", "C"},
}
``` 

I haven't tested it, but that should produce the result you need.

**Comment by StringEpsilon** [comment]
Oh wait, have I misunderstood your code? Do you want `innerArray`  to be inlined into the first array?

Cause while that is possible to make work, I'm not sure how I feel about that kind of shenanigans. Making your own intermediary representation with dictionaries isn't too hard either. But I'll think about it.

Updated example code:

```cs
var data = new Dictionary<object, object>() {
    {1, new InnerObject() },
    {2, new InnerObject() },
    {3, new InnerObject() },
    {"a", "A"},
    {"b", "B"},
    {"c", "C"},
}
```

**Comment by mrhappyasthma** [comment]
I'm just using an object for convenience, since in my case all of the keys are fixed (because I'm copying existing messages from an existing application). So your secondary example is more like what I want, except I want the object serialized like a dict.

```
object value = new InnerObject(); // just for example.

var data = new Dictionary<object, object>() {
    {1, new Dictionary<string, object>() { <key_value_pairs> },
    {2, new Dictionary<string, object>() { <key_value_pairs> },
    {3, new Dictionary<string, object>() { <key_value_pairs> },
    {"a", "A"},
    {"b", "B"},
    {"c", "C"},
}
```

It would just be nice to be able to formalize this into an object on the C# side (not serialized as an "object", but just for convenience when interacting with this in C# code).

My `a`, `b`, `c`, and all of the list values have fixed key names.

So writing it as classes in C# and just serializing each property with `PhpProperty` works nearly perfectly. I just am missing the mix of integer and string keys aspect. And was hoping to not have to write shims for each class to convert them to the representation, if possible.

**Comment by mrhappyasthma** [comment]
I suppose another option is to have our own fork with customizations to allow annotating properties as using `integer keys`. If you have any code pointers.

**Comment by StringEpsilon** [comment]
I think in that case the easiest option would be to add the capability to define integer keys via `[PhpProperty()]`. 

So you could do this:

```cs
public class MyClass {
    [PhpProperty(0)]
    public InnerObject UsefulName {get;set;}
    [PhpProperty(1)]
    public InnerObject Foo {get;set;}
    [PhpProperty(2)]
    public InnerObject Bar {get;set;}
    [PhpProperty("a")]
    public string A {get;set;}
    [PhpProperty("b")]
    public string B {get;set;}

}

**Comment by StringEpsilon** [comment]
I can make that work. Gimme an hour or so for a PoC.

**Comment by mrhappyasthma** [comment]
Yeah, exactly! If we can get something like that it'l be perfect for my needs.

**Comment by StringEpsilon** [comment]
https://github.com/StringEpsilon/PhpSerializerNET/blob/2d4c48fcb6aded904e5d7a9d0c8d109314119d09/PhpSerializerNET.Test/Serialize/ArraySerialization.cs#L23-L33

This be it?

I basically got the de-serialization for free too. If that works for you, I'll get a release out tomorrow.

**Comment by mrhappyasthma** [comment]
I cloned this and tested it and it works perfectly for my needs!

E.g. a test case along these lines (but I also tried with my real keys and values and did a `diff` to confirm a 100% match)

```
		// TODO: TEMPORARY FOR TESTING
		public class MixedKeysObjectTwo {
			[PhpProperty(0)]
			public InnerObject Foo { get; set; }
			[PhpProperty(1)]
			public InnerObject Bar { get; set; }
			[PhpProperty("a")]
			public string Baz { get; set; }
			[PhpProperty("b")]
			public string Dummy { get; set; }
		}

		public class InnerObject {
			[PhpProperty("d")]
			public string Baz { get; set; }
			[PhpProperty("e")]
			public string Dummy { get; set; }
		}
```

**Comment by StringEpsilon** [comment]
Published 1.2.0. Should be available shortly.
